### PR TITLE
chore: remove compiler warnings

### PR DESCRIPTION
**Pull request by gwhitney**
Most warnings were caused by deprecated std library templates,
  especially std::mem_fun, std::bind1st, and std::bind2nd. These
  are replaced by their modern versions, std::mem_fn and std::bind.

Note: I have just gotten a new computer and so am in the process of re-compiling and installing fluxbox. This provided the opportunity for me to look at and attempt to update/improve my several pull requests of the past. I see that the oldest (and perhaps most important one) #58 failed to be merged because I folded warning removal into the PR (as I worried about at the time I filed it, as you can see in the discussion there). So in my effort to update/revive #58, I felt it would be best first to remove the warnings in a stand-alone PR. That's what this is. Hopefully this can be merged. Once it is, I can file a new PR superseding #58 to respect _NET_WM_VISIBLE_NAME on top of a warning-free base. (My apologies, but working on the EWMH code with the templates in the signal system in the presence of many other warnings just exceeded my capacity.)

Oddly enough, in the intervening time, or perhaps just because there may be a somewhat different environment on the new computer even though it is still openSUSE Tumbleweed, there were no warnings at all about signed/unsigned mismatches, even though clearly a year or so ago I was encountering many. Hopefully that will also make this PR easier/quicker to merge, as I believe it has significantly fewer changes than the (warning parts of) #58.

Looking forward hopefully to getting through this and getting some of the features/fixes merged into fluxbox. Thanks!

### CODE REVIEW

**Comment by gwhitney** [comment]
OK, I have all of the PRs that I submitted earlier (~1 year ago) redone on top of this. So I can re-file them and close the earlier ones once this has been merged. Let me know if there is anything you'd like me to do to help with this merge. Thanks!

**Comment by spi5498** [comment]
Hello,

I have noticed the same. I see a lot of warnings for auto_ptr when compiling on Gentoo.
This has been replaced by a modern equivalent and after some research I can see that it (should) be safe to only replace this, like a search and replace deal.

I'm in no way any expert on this, however I would love to assist and put in the work on this.

Feel free to reach out in any way if my assistance is needed.

**Comment by mgumz** [comment]
@gwhitney This whole PR switches the c++-standard-compability to … which flavour? C++14? I lost touch with C++ over the course of the last decade or so … 

Another user reached out to push a little bit this PR and it looks sane for my old eyes :)

File: src/fluxbox.cc [hunk 1]
```diff
@@ -514,7 +514,7 @@ void Fluxbox::initScreen(BScreen *screen) {
     // initiate atomhandler for screen specific stuff
     STLUtil::forAll(m_atomhandler, 
             CallMemFunWithRefArg<AtomHandler, BScreen&, void>(&AtomHandler::initForScreen, *screen));
-    //STLUtil::forAll(m_atomhandler, bind2nd(mem_fun(&AtomHandler::initForScreen), *screen));
+    //STLUtil::forAll(m_atomhandler, bind2nd(mem_fn(&AtomHandler::initForScreen), *screen));
```
Comment (luebking) [inline src/fluxbox.cc:517]: stale commented bind2nd, only mem_fun replaced

**Comment by ObligedTester** [comment]
I have performed tests as late as today on this branch with successful results.

```
Fluxbox version: 1.3.7
Compiled: Jun  9 2024 15:34:02
Compiler: GCC
Compiler version: 13.2.1 20240210
```
Compiler version is the latest stable release for Gentoo on x86_64

The mentioned compiler warnings are no longer present. Fluxbox works without issue in my use-case.

As a proof read of sorts, I made my own local version and researched the necessary changes for these warnings and came to the same conclusions as @gwhitney 

Worth noting though: I do get this warning(Completely unrelated to the PR, but still a warning):

```
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/x86_64-pc-linux-gnu/bits/c++allocator.h:33,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/allocator.h:46,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/string:43,
                 from src/FbTk/RegExp.hh:29,
                 from src/ClientPattern.hh:27,
                 from src/WorkspaceCmd.hh:28,
                 from src/WorkspaceCmd.cc:23:
In member function ‘void std::__new_allocator<_Tp>::deallocate(_Tp*, size_type) [with _Tp = std::_List_node<FluxboxWindow*>]’,
    inlined from ‘static void std::allocator_traits<std::allocator<_CharT> >::deallocate(allocator_type&, pointer, size_type) [with _Tp = std::_List_node<FluxboxWindow*>]’ at /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/alloc_traits.h:517:23,
    inlined from ‘void std::__cxx11::_List_base<_Tp, _Alloc>::_M_put_node(typename _Node_alloc_traits::pointer) [with _Tp = FluxboxWindow*; _Alloc = std::allocator<FluxboxWindow*>]’ at /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/stl_list.h:522:39,
    inlined from ‘void std::__cxx11::list<_Tp, _Alloc>::_M_erase(iterator) [with _Tp = FluxboxWindow*; _Alloc = std::allocator<FluxboxWindow*>]’ at /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/stl_list.h:2024:13,
    inlined from ‘void std::__cxx11::list<_Tp, _Alloc>::pop_back() [with _Tp = FluxboxWindow*; _Alloc = std::allocator<FluxboxWindow*>]’ at /usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/stl_list.h:1341:23,
    inlined from ‘virtual void ArrangeWindowsCmd::execute()’ at src/WorkspaceCmd.cc:444:32:
/usr/lib/gcc/x86_64-pc-linux-gnu/13/include/g++-v13/bits/new_allocator.h:172:33: warning: ‘void operator delete(void*, std::size_t)’ called on unallocated object ‘normal_windows’ [-Wfree-nonheap-object]
  172 |         _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n));
      |                                 ^
src/WorkspaceCmd.cc: In member function ‘virtual void ArrangeWindowsCmd::execute()’:
src/WorkspaceCmd.cc:420:24: note: declared here
  420 |     Workspace::Windows normal_windows;
      |                        ^~~~~~~~~~~~~~
```
I do however suspect that this might be due to false positive bugs as discussed here: https://stackoverflow.com/questions/75348400/warning-operator-delete-called-on-unallocated-object
When compiling without -O2, the warning is no longer present.

Please let me know if any further testing is wanted, different compiler versions or flags and so on. 
I have access to both virtual machines and physical hardware.

**Comment by crpb** [comment]
> Fluxbox version: 1.3.7
GIT Revision: 9d8202f32338a3f08d3fa39057dc5eec5d97be4e
Compiled: Jun  9 2024 17:06:17
Compiler: GCC
Compiler version: 12.2.0

File: src/fluxbox.cc (hunk 1 above)
Comment (gwhitney) [reply src/fluxbox.cc:517]: My apologies for not fully understanding: Is this comment a request for a change to this PR? Happy to make the desired change if you can let me know what you would like. It is true that because of a search-and-replace, this change was made in a commented-out line, but it is consistent with the operative changes and didn't seem harmful, as the line is commented out anyway. Thanks for letting me know your preference here.

Comment (luebking) [reply src/fluxbox.cc:517]: I mostly wanted to sign that "yup, I've read that" and this is the only thing to add to it from here.

The "problem" isn't the change in the commented code - that's fine.
The "problem" is that you fixed the mem_fun reference but the bin2nd patter remains so if anyone ever figures to uncomment that line, and isolated warning (or by that time error) will show up here.
I'd suggest to change the bind2nd in the context this commit as weill (though inert because of the comment)

Comment (gwhitney) [reply src/fluxbox.cc]: Done, thanks for clarifying.

**Comment by gwhitney** [comment]
> @gwhitney This whole PR switches the c++-standard-compability to … which flavour?

To the default C++ standard used by `g++` on my installation of OpenSUSE Tumbleweed. That's g++ version 13.2, which defaults to C++17.

One could freeze the C++ standard by adding a `-std=`... option to the g++ options in the makefile. Would you like me to add a commit for that, and if so, which standard would you like it to be? I'm willing to go right to C++20 on the presumption that would provide the longest "shelf life"; there might be additional warning/error changes needed if so, not sure.

**Comment by mgumz** [comment]
@luebking any objections towards c++20?

**Comment by luebking** [comment]
You'd not even have to go for c++20 to immediately replace all the binds with lambdas ;)
Basically the question is whether by enforcing c++20 you'd
> there might be additional warning/error changes needed if so, not sure.

end up with a massive patch or are just telling everyone "please don't introduce old constructs".
(I'd personally not introduce arbitrary restrictions that then lead to huge patch requirements)

The other consideration is that debian buster ("oldoldstable") currently ships GCC 8 which has only limited c++20 support and would be excluded by this move. Bullseye ("oldstable") only ships GCC 10 which has c++20 support but lacks some features and bookworm ("stable") w/ GCC 12 might run into issues w/ fringe constructs (but that's a completely theoretical situation)
tl;dr: c++20 will exclude debian buster for sure and might run into occasional problems for bullseye. Whether there's any interest for future fluxbox releases by those distros (if there's a fluxbox release before the next debian cycle) I can't tell - I *guess* not?


https://gcc.gnu.org/projects/cxx-status.html#cxx20
https://packages.debian.org/search?suite=all&searchon=names&keywords=gcc

**Comment by mgumz** [comment]
@luebking you see, at least one person on earth needs to be in sync with c++ versions in some form :)

ok, so lets keep that force-to-newer-c++version out of this PR here. the less we, in the end, depend on specifics - the better.

**Comment by crpb** [comment]
https://qa.debian.org/cgi-bin/madison.cgi?package=gcc&table=debian&a=&c=&s=#

**Comment by gwhitney** [comment]
> ok, so lets keep that force-to-newer-c++version out of this PR here. the less we, in the end, depend on specifics - the better.

Yes, totally your call. But would you like me to introduce a `-std=c++17` option in make to be specific so that in the future when (let's suppose) C++20 becomes the default, it won't introduce new warnings? Or, conversely, drop this PR and replace it with one that adds a `-std=` option with an older C++ standard that will perhaps make the existing code compile without warnings, if possible? Just want to implement what you feel is best for FluxBox so that hopefully it's possible to move forward with some of the other constructive PRs. Thanks for letting me know.

**Comment by mgumz** [comment]
@gwhitney lets finish with this one here for now and not going into `--std` land yet.

**Comment by gwhitney** [comment]
> @gwhitney lets finish with this one here for now and not going into `--std` land yet.

Absolutely fine. In that case, I am not aware of anything else waiting on me for the sake of this PR -- if there is anything else you'd like me to do, please just let me know. Thanks!

**Comment by gwhitney** [comment]
Given that this PR recently celebrated its second birthday, and there have been two commits to this repository in the past 28 months, I no longer have any expectation of continued maintenance/development of fluxbox (alas).

**Comment by mgumz** [comment]
@gwhitney so you want to trigger me :)

**Comment by gwhitney** [comment]
> @gwhitney so you want to trigger me :)

Not at all. The fact that I am still using fluxbox years after it's gone pretty quiescent is testament to my tremendous respect for what you (and presumably some colleagues of yours, I am not familiar with the exact history) created.

Somewhere along the history of user-interface development, the world took an unfortunate turn in which each different application develops its own idiosyncratic tabbing mechanism, when it's clear that such a mechanism should be created at the window/app manager level, so that tabbing will have uniform appearance, operate uniformly, and will achieve the power and convenience of mixed-app tab groups. Yet to the best of my knowledge, only your fluxbox ever achieved that excellent design with a very comfortable, usable interface. I don't know of any other window manager with nearly as well-developed of a tabbing feature.

Anything you may have read into my comment is just a trace of disappointment, along with trying to realistically set the expectations of others who may visit this GitHub repo.

I applaud you.

**Comment by mgumz** [comment]
> It may be helpful to document to which stdc++ fluxbox refers to, in some distinct file. Right now external contributors may not be certain which stdc++ applies. Having this in a single file, be it a markdown file or anything else, may clarify this for future contributors too. In this PR alone and the last one I looked at, in another pull request, we have at the least mentioned C++14, C++17 and C++20. Soon it'll be the year 2025, so ...

@rubyFeedback feel free to provide one as you see fit.

**Comment by mgumz** [comment]
@gwhitney the "pause" here (or the other PRs) was/is not intentional. as you said: your pr was … ready. just be more … pushy, so to speak. out-of-fluxbox events push silent PRs aside.

please do not take it personally, it was not meant as such.
